### PR TITLE
PLT-8744 Introduce ExtLedgerStateCoordinator

### PR DESCRIPTION
**Pull request by berewt**
Many indexers need access to `ExtLedgerState` to get their indexing information. As maintaining the ledgerstate is expensive, what we did so far was hardcoding the `EpochStateIndexer` that was both maintaining the ledger state and forwarding it to other indexers.
But indexer were hardcoded in the `EpochStateIndexer`, and it limited its reusability.

This PR introduce the `ExtLedgerStateCoordinator`, which acts as a coordinator while maintaining the `ExtLedgerState`. This way, one can reuse it with its desired set of workers.
 

Pre-submit checklist:
- Branch
    - [ ] Tests are provided (if possible)
    - [ ] Commit sequence broadly makes sense and have useful messages
    - [ ] Important changes are reflected in changelog.d of the affected packages
    - [ ] Relevant tickets are mentioned in commit messages
- PR
    - [ ] (For external contributions) Corresponding issue exists and is linked in the description
    - [ ] Targeting main unless this is a cherry-pick backport
    - [ ] Self-reviewed the diff
    - [ ] Useful pull request description
    - [ ] If relevant, reference the ADR in the PR and reference the PR in the ADR
    - [ ] Reviewer requested

### CODE REVIEW

File: marconi-chain-index/src/Marconi/ChainIndex/Indexers/ExtLedgerStateCoordinator.hs [hunk 1]
```diff
@@ -0,0 +1,757 @@
...
+  }
+
+{-
+  p <- max <$> Core.lastStablePoint epochStateIndexer' <*> Core.lastStablePoint epochBlocksIndexer
+  ledgerState <-
+    either
+      (const $ throwError $ Core.IndexerInternalError "Can't resolve the last stable point")
+      pure
+      =<< runExceptT (restoreLedgerStateFromIndexers genesisCfg p epochStateIndexer' epochBlocksIndexer)
+-}
```
Comment (ana-pantilie) [inline marconi-chain-index/src/Marconi/ChainIndex/Indexers/ExtLedgerStateCoordinator.hs:234]: Should this be removed?

**Review by ana-pantilie — COMMENTED** [review]
I have some questions, I need to understand them first before I can review the rest.

File: marconi-chain-index/src/Marconi/ChainIndex/Indexers/ExtLedgerStateCoordinator.hs [hunk 2]
```diff
@@ -0,0 +1,757 @@
...
+import Ouroboros.Consensus.Node.NetworkProtocolVersion qualified as O
+import Ouroboros.Consensus.Node.Serialisation qualified as O
+import Ouroboros.Consensus.Shelley.Ledger qualified as O
+import Ouroboros.Consensus.Storage.Serialisation qualified as O
+import Prettyprinter (Pretty (pretty))
+import System.Directory (createDirectoryIfMissing)
+import System.FilePath ((</>))
+import Text.Read qualified as Text
+
+type ExtLedgerState = O.ExtLedgerState (O.HardForkBlock (O.CardanoEras O.StandardCrypto))
```
Comment (ana-pantilie) [inline marconi-chain-index/src/Marconi/ChainIndex/Indexers/ExtLedgerStateCoordinator.hs:81]: Oh this is where the EUTxOs are? So the modelling of this data is in the Ouroboros repository, right?

File: marconi-chain-index/src/Marconi/ChainIndex/Indexers.hs [hunk 3]
```diff
@@ -354,29 +366,62 @@ mintBuilder securityParam catchupConfig mintEventConfig textLogger path =
           (BM.appendName indexerName indexerEventLogger)
    in MintTokenEvent.mkMintTokenEventWorker mintTokenWorkerConfig mintEventConfig mintDbPath
 
--- | Configure and start the @EpochState@ indexer
-epochStateBuilder
-  :: (MonadIO n, MonadError Core.IndexerError n)
+-- | Configure and start the @EpochNonce@ indexer
+epochNonceBuilder
+  :: (MonadIO n, MonadError Core.IndexerError n, MonadIO m)
```
[the remark this answers is not in the document]
Comment (berewt) [reply marconi-chain-index/src/Marconi/ChainIndex/Indexers.hs:371]: No specific need, it just happened a few time in the past that a strong commitment to IO came in our way so I changed it.

File: marconi-chain-index/src/Marconi/ChainIndex/Indexers.hs [hunk 4]
```diff
@@ -354,29 +366,62 @@ mintBuilder securityParam catchupConfig mintEventConfig textLogger path =
...
   => SecurityParam
   -> Core.CatchupConfig
-  -> EpochState.EpochStateWorkerConfig
-  -> BM.Trace IO Text
+  -> BM.Trace m Text
```
[the remark this answers is not in the document]
Comment (berewt) [reply marconi-chain-index/src/Marconi/ChainIndex/Indexers.hs:374]: I can try to remove while I'm at it, if I can't, I'll open a ticket.

File: marconi-chain-index/src/Marconi/ChainIndex/Indexers/EpochSDD.hs [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+
+-- | Event for @SDD@ storage
+data EpochSDD = EpochSDD
+  { _sddEpochNo :: !C.EpochNo
+  , _sddPoolId :: !C.PoolId
+  , _sddLovelace :: !C.Lovelace
+  , _sddBlockNo :: !C.BlockNo
+  }
+  deriving stock (Eq, Ord, Show, Generic)
+  deriving anyclass (SQL.FromRow, SQL.ToRow, FromJSON, ToJSON)
```
[the remark this answers is not in the document]
Comment (berewt) [reply marconi-chain-index/src/Marconi/ChainIndex/Indexers/EpochSDD.hs:76]: LedgerState contains (amongst other thing, the active set of eUTxOs). The `EpochSDD` contains the stakepool distribution (the reward offered to each stakepool for their contribution to the network).

File: marconi-chain-index/src/Marconi/ChainIndex/Indexers/ExtLedgerStateCoordinator.hs (hunk 1 above)
Comment (berewt) [reply marconi-chain-index/src/Marconi/ChainIndex/Indexers/ExtLedgerStateCoordinator.hs:234]: Indeed, legacy of the last changes I made to the module.

File: marconi-chain-index/src/Marconi/ChainIndex/Indexers/ExtLedgerStateCoordinator.hs (hunk 2 above)
Comment (berewt) [reply marconi-chain-index/src/Marconi/ChainIndex/Indexers/ExtLedgerStateCoordinator.hs:81]: Yes. :)

File: marconi-chain-index/src/Marconi/ChainIndex/Indexers.hs (hunk 4 above)
[the remark this answers is not in the document]
Comment (berewt) [reply marconi-chain-index/src/Marconi/ChainIndex/Indexers.hs:374]: New ticket, I'll create it.

**Review by ana-pantilie — APPROVED** [review]

**Pull request merged**